cmd/whitesourceExecuteScan.go

```diff
@@ -291,9 +574,125 @@ func writeWhitesourceConfigJSON(config *ScanOptions, utils whitesourceUtils, dev
+func executeNpmScan(config *ScanOptions, scan *whitesourceScan, utils whitesourceUtils) error {
+	modules, err := utils.FindPackageJSONFiles(config)
+	if err != nil {
+		return fmt.Errorf("failed to find package.json files with excludes: %w", err)
+	}
+	if len(modules) == 0 {
+		return fmt.Errorf("found no NPM modules to scan. Configured excludes: %v",
+			config.BuildDescriptorExcludeList)
+	}
+	for _, module := range modules {
+		err := executeNpmScanForModule(module, config, scan, utils)
+		if err != nil {
+			return fmt.Errorf("failed to scan NPM module '%s': %w", module, err)
+		}
+	}
+	return nil
+}
+
+// executeNpmScanForModule generates a configuration file whitesource.config.json with appropriate values from config,
+// installs all dependencies if necessary, and executes the scan via "npx whitesource run".
+func executeNpmScanForModule(modulePath string, config *ScanOptions, scan *whitesourceScan, utils whitesourceUtils) error {
+	log.Entry().Infof("Executing Whitesource scan for NPM module '%s'", modulePath)
+
+	resetDir, err := utils.Getwd()
+	if err != nil {
+		return fmt.Errorf("failed to obtain current directory: %w", err)
+	}
+
+	dir := filepath.Dir(modulePath)
+	if err := utils.Chdir(dir); err != nil {
+		return fmt.Errorf("failed to change into directory '%s': %w", dir, err)
+	}
+	defer func() {
+		err = utils.Chdir(resetDir)
+		if err != nil {
+			log.Entry().Errorf("Failed to reset into directory '%s': %v", resetDir, err)
+		}
+	}()
+
+	if err := writeWhitesourceConfigJSON(config, utils, false, true); err != nil {
+		return err
+	}
+	defer func() { _ = utils.FileRemove(whiteSourceConfig) }()
+
+	projectName, err := getNpmProjectName(modulePath, utils)
+	if err != nil {
+		return err
+	}
+
+	if err := reinstallNodeModulesIfLsFails(modulePath, config, utils); err != nil {
+		return err
+	}
+
+	if err := scan.appendScannedProject(projectName); err != nil {
+		return err
+	}
+
+	return utils.RunExecutable("npx", "whitesource", "run")
+}
+
+func getNpmProjectName(modulePath string, utils whitesourceUtils) (string, error) {
+	fileContents, err := utils.FileRead("package.json")
+	if err != nil {
+		return "", fmt.Errorf("could not read package.json: %w", err)
+	}
+	var packageJSON = make(map[string]interface{})
+	err = json.Unmarshal(fileContents, &packageJSON)
+
+	projectNameEntry, exists := packageJSON["name"]
+	if !exists {
+		return "", fmt.Errorf("the file '%s' must configure a name",
+			filepath.Join(modulePath, "package.json"))
+	}
+
+	projectName, isString := projectNameEntry.(string)
+	if !isString {
+		return "", fmt.Errorf("the file '%s' must configure a name",
+			filepath.Join(modulePath, "package.json"))
+	}
+
+	return projectName, nil
+}
+
+func reinstallNodeModulesIfLsFails(modulePath string, config *ScanOptions, utils whitesourceUtils) error {
+	// No need to have output from "npm ls" in the log
+	utils.Stdout(ioutil.Discard)
+	defer utils.Stdout(log.Writer())
+
+	err := utils.RunExecutable("npm", "ls")
```

Comment:
do we need this?
I think WhiteSource has an option to ignore `ls` errors but maybe I overlook s.th.

Reply:
You could be right, but I assumed the Cloud SDK implementation included this for a reason. I have not challenged whether this is valid or not, since I was not sure I could reproduce problems in my test projects.

Reply:
Unfortunately the npm ls ignore option did not work properly as far as I remember

Reply:
But maybe this has changed or was influenced by another problem.